tests/processing/strategies/test_all.py

```diff
@@ -0,0 +1,189 @@
+    next_step: DummyStrategy, raises_invalid_message: bool = False
+) -> FilterStep[bool]:
+    if raises_invalid_message:
+        pytest.skip("does not support invalid message")
+    return FilterStep(
+        lambda message: message.payload, next_step, commit_policy=IMMEDIATE
+    )
+
+
+def reduce_step_factory(
+    next_step: DummyStrategy, raises_invalid_message: bool = False
+) -> Reduce[bool, bool]:
+    if raises_invalid_message:
+        pytest.skip("does not support invalid message")
+
+    def accumulator(result: bool, value: BaseValue[bool]) -> bool:
+        assert isinstance(result, bool)
+        assert isinstance(value.payload, bool)
+        return value.payload
+
+    return Reduce(
+        max_batch_size=1,
+        max_batch_time=1,
+        accumulator=accumulator,
+        initial_value=bool,
+        next_step=next_step,
+    )
+
+
+FACTORIES: Sequence[StrategyFactory] = [
+    run_task_factory,
+    run_task_with_multiprocessing_factory,
+    run_task_in_threads_factory,
+    filter_step_factory,
+    reduce_step_factory,
+]
+
+
+@pytest.mark.parametrize("strategy_factory", FACTORIES)
+def test_filters(strategy_factory: StrategyFactory) -> None:
+    next_step = Mock()
+    step = strategy_factory(next_step)
+
+    def test_function(message: Message[bool]) -> bool:
+        return message.payload
+
+    messages: Sequence[DummyMessage] = [
+        Message(Value(True, {Partition(Topic("topic"), 0): 1})),
+        Message(Value(FILTERED_PAYLOAD, {Partition(Topic("topic"), 0): 2})),
+        Message(Value(True, {Partition(Topic("topic"), 0): 3})),
+    ]
+
+    for message in messages:
+        step.poll()
+        step.submit(message)
+
+    step.close()
+    step.join()
+
+    assert next_step.submit.call_args_list == list(map(call, messages))
+
+
+@pytest.mark.parametrize("strategy_factory", FACTORIES)
+def test_dlq(strategy_factory: StrategyFactory) -> None:
+    next_step = Mock()
+    step = strategy_factory(next_step, raises_invalid_message=True)
+
+    def test_function(message: Message[bool]) -> bool:
+        return message.payload
+
+    messages = [
+        Message(Value(True, {Partition(Topic("topic"), 0): 1})),
+        Message(Value(False, {Partition(Topic("topic"), 0): 2})),
+        Message(Value(True, {Partition(Topic("topic"), 0): 3})),
+    ]
+
+    step.poll()
+    step.submit(messages[0])
+
+    invalid_messages = []
+
+    step.poll()
+    try:
+        step.submit(messages[1])
+    except InvalidMessage as e:
+        invalid_messages.append(e)
+
+    try:
+        step.poll()
+    except InvalidMessage as e:
+        invalid_messages.append(e)
+
+    step.submit(messages[2])
+
+    step.close()
+
+    try:
+        step.join()
+    except InvalidMessage as e:
+        invalid_messages.append(e)
+
+    assert len(invalid_messages) == 1
+
+    assert next_step.submit.call_args_list == [
+        call(messages[0]),
+        call(Message(Value(FILTERED_PAYLOAD, {Partition(Topic("topic"), 0): 2}))),
```

Comment:
I think we don't actually want this one to be submitted from every strategy. Otherwise an InvalidMessage can generate way too many filtered messages as it gets propagated through a sequence of processing steps. Only certain strategies should send it like the decoder or run task (if it is raised by the processing function passed in by the user)

Reply:
yeah that's fine. what matters is that the unfiltered messages come in the right order and that the merged committable of all messages is correct